src/Features/Core/Portable/ConvertForToForEach/AbstractConvertForToForEachCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,489 @@
+                forStatement,
+                (currentFor, _) => this.ConvertForNode(
+                    (TForStatementSyntax)currentFor, typeNode, foreachIdentifier,
+                    collectionExpression, iterationType, options));
+
+            return document.WithSyntaxRoot(editor.GetChangedRoot());
+
+            // local functions
+            (TTypeNode, SyntaxToken, TStatementSyntax) tryDeconstructInitialDeclaration()
+            {
+                if (bodyStatements.Count >= 1)
+                {
+                    var firstStatement = bodyStatements[0];
+                    if (syntaxFacts.IsLocalDeclarationStatement(firstStatement))
+                    {
+                        var variables = syntaxFacts.GetVariablesOfLocalDeclarationStatement(firstStatement);
+                        if (variables.Count == 1)
+                        {
+                            var firstVariable = (TVariableDeclaratorSyntax)variables[0];
+                            if (IsValidVariableDeclarator(firstVariable))
+                            {
+                                var firstVariableInitializer = syntaxFacts.GetValueOfEqualsValueClause(
+                                    syntaxFacts.GetInitializerOfVariableDeclarator(firstVariable));
+                                if (syntaxFacts.AreEquivalent(firstVariableInitializer, indexExpression))
+                                {
+                                    var type = (TTypeNode)syntaxFacts.GetTypeOfVariableDeclarator(firstVariable)?.WithoutLeadingTrivia();
+                                    var identifier = syntaxFacts.GetIdentifierOfVariableDeclarator(firstVariable);
+                                    var statement = firstStatement;
+                                    return (type, identifier, statement);
+                                }
+                            }
+                        }
+                    }
+                }
+
+                return default;
+            }
+
+            void findAndReplaceMatches(SyntaxNode current)
+            {
+                if (syntaxFacts.AreEquivalent(current, indexExpression))
+                {
+                    // Found a match.  replace with iteration variable.
+                    var replacementToken = foreachIdentifierReference;
+
+                    if (semanticFacts.IsWrittenTo(semanticModel, current, cancellationToken))
+                    {
+                        replacementToken = replacementToken.WithAdditionalAnnotations(
+                            WarningAnnotation.Create(FeaturesResources.Warning_colon_Collection_was_modified_during_iteration));
+                    }
+
+                    if (crossesFunctionBoundary(current))
+                    {
+                        replacementToken = replacementToken.WithAdditionalAnnotations(
+                            WarningAnnotation.Create(FeaturesResources.Warning_colon_Iteration_variable_crossed_function_boundary));
```

Comment:
Was this covered in a test?

Reply:
yes.  "TestWarnIfCrossesFunctionBoundary"

Reply:
#resolved